proposals/0000-allow-constraint-type-synonyms-in-deriving-heads.md

```diff
@@ -0,0 +1,200 @@
+2. For any constraint tuple in derived instance head,
+   split such declarations into multiple.
+
+After that rewrites happen, existing GHC semantics is applied to result,
+with same error messages or produced code.
+
+There are two cases not covered by this rewrites:
+
+1. Using tuples in instance context leads to nested tuples.
+   But they already have the same semantics as flatten tuples,
+   so this is concern only affecting AST handling.
+2. Syntax and need for user-defined instances of class tuples is not clear.
+   So tuples in user-defined instances head remain forbidden.
+
+## Examples
+
+Main usecase:
+
+```haskell
+type ToFromJSON x = (ToJSON x, FromJSON x)
+
+-- Allowed by this proposal
+derive instance ... => (ToFromJSON x)
+
+-- Should work the same with syntax allowed before proposal
+derive instance ... => ToJSON x
+derive instance ... => FromJSON x
+```
+
+Instance context case is simpler:
+
+```haskell
+class ToFromJSON x => MyProtocolEncoding x where
+   ...
+
+data MyContainer x = ...
+
+-- Allowed by this proposal
+instance
+   (MyProtocolEncoding x, ToFromJSON (MyContainer x)) =>
+   MyContainer x where
+   ...
+
+-- Should work the same with syntax allowed before proposal
+instance
+   (MyProtocolEncoding x, ToJSON (MyContainer x), FromJSON (MyContainer x)) =>
+   MyContainer x where
+   ...
+```
+
+Last example may be not the most natural,
+because `MyContainer` would probably have `ToFromJSON` conditional instance
+anyway. But my goal was to show instance context changes separately,
+while more natural examples involve instance head as well.
+
+Non-standalone instances do not have explicit params,
+so rewriting implementation may be little different.
+But they have just the same samantics as standalone instances with holes,
+covered by examples before, so could be handled the same.
+
+## Effect and Interactions
+
+There is already extension named `TypeSynonymInstances`,
+but it does not cover constraint synonym case.
+Probably, constraint synonyms should be guarded by this extension as well
+and documentation updated on that.
+
+It may be, that some linters depend on current syntaxic restrictions,
+but I am not aware on any specific cases.
+
+## Costs and Drawbacks
+
+Main concern is that rewriting implementations may make errors less clear.
+I propose to mitigate if this happens by adding context to errors,
+same as it done for instances resolving now.
+
+## Backward Compatibility
+
+No breakage. Part of this even was working before.
+
+## Alternatives
+
+For deriving instances there are two alternatives:
+
+1. Use CPP preprocessor variable, which breaks type-safety,
+   may lead to surprising error messages and is less composable
+2. Use Template Haskell, which have its own problems and is less simple for beginners
+
+For user-defined instances one may define type class instead of synonym:
+
+```haskellb
+class (ToJSON x, FromJSON x) => ToFromJSON x
+
+instance (ToJSON x, FromJSON x) => ToFromJSON x
+```
```

Comment:
I don't think this is actually an alternative. I can't derive or define an instance of `ToFromJSON` without also (manually) deriving or defining `ToJSON` and `FromJSON` instances.

Reply:
```suggestion
```haskell
class (ToJSON x, FromJSON x) => ToFromJSON x

instance (ToJSON x, FromJSON x) => ToFromJSON x
```
```